articles/055_ros_parameter_design.md

```diff
@@ -0,0 +1,184 @@
+---
+layout: default
+title: Parameter API design in ROS
+permalink: articles/ros_parameters.html
+abstract:
+  This article is proposed design for the interfaces for interacting with parameters in ROS 2.0.
+  We focus here on specifying the system design and leave the implementation unspecified.
+author: '[Tully Foote](https://github.com/tfoote)'
+published: true
+---
+
+* This will become a table of contents (this text will be scraped).
+{:toc}
+
+# {{ page.title }}
+
+<div class="abstract" markdown="1">
+{{ page.abstract }}
+</div>
+
+Original Author: {{ page.author }}
+
+## Background
+
+In ROS 1 the parameters were implemented in a 'blackboard model' with unrestricted read and write access from all nodes.
+The data model proved useful in many cases but there were many cases where the lack of control or ownership proved to be a problem.
+One of the common shortcomings was for setting parameters on drivers.
+A tool called dynamic_reconfigure was developed to service this use case.
+It provided a service based interface to interact with parameters of other nodes.
+
+### Other resources
+
+Other resources related to the parameter design process for ROS 2.0 include:
+
+- Gonzalo's research on parameters.
+  - Discussion: https://groups.google.com/forum/#!topic/ros-sig-ng-ros/YzCmoIsN0o8 and https://groups.google.com/forum/#!searchin/ros-sig-ng-ros/parameter/ros-sig-ng-ros/fwDBcei5Ths/L6ORPfjUDXYJ
+  - Protype: https://github.com/abellagonzalo/dynamic_config
+  - Final Notes:  [http://wiki.ros.org/sig/NextGenerationROS/Parameters](http://wiki.ros.org/sig/NextGenerationROS/Parameters)
+- Thibault's nodeparam draft REP: https://github.com/tkruse/rep/blob/nodeparam/nodeparam-REP.rst
+
+## Ideal System
+
+It is useful to consider the ideal system to understand how it relates to the current system and how a new system could work.
+We'd like to support both the use cases of standard parameters as well as dynamic parameters.
+An ideal parameter system would have the qualities laid out in the following paragraphs.
+
+### Accept parameter values
+
+At the most basic a parameter system must be able to accept the setting of parameter values.
+An extension is to accept a group of parameters updated atomically.
+This allows you to update coupled parameters such as PID gains without worring about invalid intermediate states.
+
+### Return parameter values
+
+Also at the core functionality return the value of one or more parameters which have previously been set.
+
+### Unset a parameter value
+
+If a parameter has been set it should be able to be unset.
+
+### Introspection of available parameters
+
+Provide a list of currently set parameters.
+
+### Notifications of change
+
+When a parameter changes value it should be possible to get a notification.
+
+### Reject parameter changes
```

Comment:
If I understand correctly, this would be the method for enforcing parameter ranges (among other things). And 'be able to communicate the preconditions' would refers to a mechanism that could be used to, for instance, display valid parameter values or ranges, as `rqt_reconfigure` currently does. Is my interpretation correct?